### ISSUE TITLE
Use language daemons for language providers

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#6843]
It is not unknown that `:python` and other languages startup is being slow now due to having to check and start external programs. E.g. with `init.vim` as simple as

```VimL
" let g:powerline_old_vim = 1
set laststatus=2
python import vim
python from powerline.vim import setup as powerline_setup
python powerline_setup(gvars=globals())
python del powerline_setup

set guicursor=
```

(using `editors-refactor` branch of powerline, powerline/powerline#1507) Vim takes 54 ms to source that, Neovim takes 325 ms, all with system FS caches up, worse without. The problem is that each Neovim has to start Python by itself.

I have a suggestion thus: instead of having architecture with which one Neovim has one Python which it directly communicates with add a “language daemon”. The idea is that “language daemon” starts once and creates a socket in a set location used to communicate with Neovim, when Neovim needs `:python` it starts a daemon if it was not already started (decreased performance, but only at startup, and “starting a daemon” may be a per-user systemd service), then on `:python` it just tells daemon that Neovim client with ID X (Neovim socket, but generally just any unique ID) needs to run `:python` command Y. If such ID is unknown to a language server then it spawns a new Python child process and relay requested command there, if it is known it either relays requested command or gives some details regarding the fate of the child (better error reporting if Python crashed).

Additionally if relaying does not work well in term of performance it should be possible to use language server to communicate direct interaction, then in “channel is broken” cases it should also be possible to get some details from the language server.

The main ideas here are

1. `fork()` is much cheaper then `fork()+exec()`.
2. Need not run checks regarding the availability and version of `neovim` packages more then once.

Main problems:

1. To update `neovim` package one will need to restart daemon.
2. Handling virtualenvs is going to be complicated.
3. Actually Neovim can’t just “request daemon to run command Y”, need to transfer environment variables at first command.

Both problems may be solved by another complication: make daemon a daemon manager: daemon manager is communicating with Neovim like daemon in the suggestion, but instead it spawns daemons  (key used to determine whether new daemon is needed: periodically checked `neovim` package version plus significant environment variables like where virtualenv is defined, or path to Python executable (`$PATH`, but only if it really changes executable file location), or Python package location (`$PYTHONPATH`); insignificant may just be changed on language process startup) using `fork()+exec()` (i.e. `subprocess`), daemons themselves only `fork()` to provide language processes.

Note that while my suggestion may bring down startup times significantly and provide better error reporting, this complicates code a lot and there are different solutions possible for error reporting in case of crashed language process.

**Comment by ZyX-I** [comment]
Some addition: if using “language daemon manager” it may and should be written in VimL+lua, this will also speed up failing `has('language')` checks.

**Comment by justinmk** [comment]
There are still some simple things we haven't done, that we should try, before doing something complicated.

- `has('foo')` can be near-instantaneous by just caching the result. The _worst_ thing that can happen is `:python` fails, and user has to restart because some plugin used the outdated cached value. 
    - Use confusion in this case can be mitigated by a clear error message.
- We can save ~50ms in startup cost via "eager" checks using `jobstart()`

Other notes:

- Even with daemons, milliseconds-obsessed users will notice a discrepancy vs Vim, in some cases, because of RPC vs FFI. So there will still be complaints, despite having an elaborate daemon backend.
- Users that care about _any_ of this should switch to Lua plugins. We should not spin our wheels worrying about a 200ms one-time cost that is _totally avoidable_ to people who really care that much.
- In general I think multi-tenancy is a very bad idea. Right now we have a simple model: one thing per Nvim instance. It would be a shame if we lose simplicity for very little gain.

**Comment by bfredl** [comment]
> We can save ~50ms in startup cost via "eager" checks using jobstart()

Yes, it would be good to be able to hide the latency in init.vim by requesting the hosts early (say if one knows one needs legacy py2 and rplugin py3), but leave code that actually interacts with the hosts until later in init.vim. The initialization of hosts need to be made more async, hopefully ChannelInfo (rather than blocking "poll" request) will help with that. Ideally even a MyPluginAsyncCommand will be made async with loading of the host.

**Comment by ZyX-I** [comment]
The lua plugins are not always an option. Additionally it is a question whether you could make checks async: starting async checks from the vimrc is *too late* to make any sense (in my example powerline needs `:python` right there), starting it before means loosing some user settings (like `g:python_host_prog`) thus forcing them to move to environment variables.

One good thing with my variant is that no changes to C code are needed, you can even create a third-party plugin with the implementation of my idea. Though after I imagined the idea I also had a feeling that it is overly complicated (and I have a few seconds *Vim* startup time now, so when I switch to Neovim additional half a second will not bother me much), but still posted it here in case somebody will think my idea is actually good enough to implement.

**Comment by ZyX-I** [comment]
Actually with my complicated setup it takes one or two seconds for Vim to start assuming FS caches (or whatever caches, but I have no idea what “whatever” is) are up (most of time they are not for unknown reason). Fourteen to twenty seconds in both (Vim and Neovim) cases if they are not, would never notice provider startup under these conditions.

**Comment by bfredl** [comment]
An alternative is autocommands that fire whenever the host is ready. This will likely fall out naturally from the host initialization / UpdateRemotePlugins refactor, which aims to make everything event driven rather than doing such things as `rpcrequest()` to different host in a loop...

**Comment by torpak** [comment]
How about specifying a way for plugins to declare what they need and only start them once it's available.
That would even allow plugins that don't depend on each other to start simultaneously.

Plugins that dont support this mechanism could be loaded before or after that.

That could speed up startup times dramatically especially when combined with on demand loading.

When most of your plugins are only needed when editing prolog code they and all there dependencies would not be needed when doing something else
